src/cace/amm/range.h

```diff
@@ -0,0 +1,186 @@
+        };                                                                                               \
+    }                                                                                                    \
+    static inline void M_C(intvl, _set_singleton)(M_C(intvl, _t) * obj, type val)                        \
+    {                                                                                                    \
+        *obj = (M_C(intvl, _t)) {                                                                        \
+            .has_min = true,                                                                             \
+            .i_min   = val,                                                                              \
+            .has_max = true,                                                                             \
+            .i_max   = val,                                                                              \
+        };                                                                                               \
+    }                                                                                                    \
+    static inline void M_C(intvl, _set_finite)(M_C(intvl, _t) * obj, type min_val, type max_val)         \
+    {                                                                                                    \
+        *obj = (M_C(intvl, _t)) {                                                                        \
+            .has_min = true,                                                                             \
+            .i_min   = min_val,                                                                          \
+            .has_max = true,                                                                             \
+            .i_max   = max_val,                                                                          \
+        };                                                                                               \
+    }                                                                                                    \
+                                                                                                         \
+    static inline bool M_C(intvl, _is_infinite)(const M_C(intvl, _t) * obj)                              \
+    {                                                                                                    \
+        return !(obj->has_min) && !(obj->has_max);                                                       \
+    }                                                                                                    \
+                                                                                                         \
+    static inline int M_C(intvl, _cmp)(const M_C(intvl, _t) * lt, const M_C(intvl, _t) * rt)             \
```

Comment:
**Comment**: The logic in this function is a bit tricky, it's unfortunate that the define prevents adding more detailed comments.

Reply:
Yes, this is a limitation that applies to M*LIB itself also. There still should be extracted doxygen `@struct` level documentation for the macro uses though. So it should be as readable/usable as upstream M*LIB docs.